docs/enhancements/cluster-deployment-condition-refactor.md

```diff
@@ -0,0 +1,96 @@
+AuthenticationFailure
+ClusterInstallCompleted
+ClusterInstallFailed
+ClusterInstallRequirementsMet
+ClusterInstallStopped
+DNSNotReady
+DeprovisionLaunchError
+IngressCertificateNotFound
+InstallImagesNotResolved
+InstallLaunchError
+InstallerImageResolutionFailed
+ProvisionFailed
+ProvisionStopped
+
+## Proposal
+
+Focusing only on the 12 conditions controlled by the clusterdeployment controller:
+
+ClusterImageSetNotFound - Move to RequirementsMet or possibly ClusterInstallRequirementsMet.
+AuthenticationFailure - Move to RequirementsMet or ClusterInstallRequirementsMet.
+ClusterInstallCompleted - Merge into Installed.
+ClusterInstallFailed - Merge into ProvisionFailed.
+ClusterInstallRequirementsMet - Merge into RequirementsMet.
+ClusterInstallStopped - Merge into ProvisionStopped.
+DNSNotReady - Move to RequirementsMet
+DeprovisionLaunchError -  Rename to DeprovisionError as a general bucket for any deprovision problems we can encounter.
+InstallLaunchError - Move to RequirementsMet or ProvisionFailed
+InstallerImageResolutionFailed - Move to RequirementsMet or ProvisionFailed
+ProvisionFailed - Keep as a primary condition.
+ProvisionStopped - Keep as a primary condition.
+
+Implied here is a new utility to check if a grouped condition reason is ours, and if we know we have now cleared that state, switch the condition back to false.
+
+
+### RequirementsMet
+
+Unknown until we're ready to launch an install, then True. False if we've found a problem.
+
+Can be set via clusterdeployment_controller up until we hand over to a ClusterInstall, or when we detect a RequirementsMet condition on the ClusterInstall. Would need to ensure that the code that copies from ClusterInstall -> ClusterDeployment only runs after we've cleared all our possible Reasons for this condition to be False in the clusterdeployment controller.
+
+Possible Reasons when False from clusterdeployment_controller:
+  - ClusterImageSetNotFound
+  - AuthenticationFailure
+  - DNSNotReady
+  - InstallerImageResolutionFailed
+  - InstallLaunchError
+
+TODO: should RequirementsMet status=False reasons also translate to a ProvisionFailed = True with the same Reason?
+
+Each of the above should use a utility to ensure they are not the reason the condition is False, once we know that problem is resolved. (set status back to Unknown)
+
+ProvisionStopped - True if we've given up. Normally should just be set via the ClusterInstall's ProvisionStopped condition.
+
+### ProvisionFailed
+
+Unknown until we hit our first error, true as soon as we detect any problem, false once we successfully install.
+
+TODO: per above, should we set this for preflight'ish requirements met errors? Would give callers one place to watch to know something is wrong.
+
+### ProvisionStopped
+
+True when the ClusterInstall implementation has given up. (via Stopped condition on ClusterInstall)
+
+### Provisioned
+
+Net new condition, true once cd.Spec.Installed = true. Allows callers to oc wait.
```

Comment:
Could we further consolidate these into just `Provisioned`?
- `status: "True"` when we're all good
- `status: "False"` otherwise, where `reason` is a breadcrumb to the Condition that provides more detail, e.g.
```yaml
    conditions:
    - message: Pre-flight requirements not met. See the RequirementsMet condition for details.
      reason: RequirementsNotMet
      status: "False"
      type: Provisioned
```
or
```yaml
    conditions:
    - message: We gave up. See the Stopped condition on ClusterInstall for details.
      reason: Stopped
      status: "False"
      type: Provisioned
```
etc.

Reply:
I think separate provides value in that you can monitor ProvisionFailed to know something is up, but it might still clear up, and ProvisionStopped meaning we've given up and there's definitely a problem. Also important to be able to see details on what failed, once you're stopped. We could redirect to the ClusterInstall for those details but that's worse UX than you get today.

Reply:
What does such monitoring look like? Why is it easier to watch for
- `type=="ProvisionStopped" && status=="True"` than `type="Provisioned" && reason=="Stopped"`
- `type=="ProvisionFailed && status=="True"` than `type="Provisioned" && reason=="Failed"`

> see details on what failed, once you're stopped

Meaning keeping the details in the message? Cool :+1:

Reply:
I feel like keeping them separate provides a bit more clear state for a couple reasons. It seems odd to react to conditions without checking the Status, and in the above that would not be possible, Provisioned would be False during install normally, so you'd have to skip that and watch for reasons. This seems less clear to code against than reacting to a condition that indicates a problem going True.

There also might be programatic code you want to react to specific situations, but if we consume Reasons for very generic situations that is no longer rpossible, whereas you could have code monitoring for ProvisionFailed=True + Reason=FooBar and something could take action to fix FooBar.

Reply:
Ight.

I'm just coming from a place where the `conditions` array has an overwhelming number of items, and the knowledge of which one(s) to look at (as a human) in what order of precedence seems tribal.

But IIUC this proposal would get us from 20-something top-level items down to four? That should be significantly less confusing -- though I would still want to make sure there are docs that explain how to read them.

Reply:
It would take ~13 of them down to 4, the other half dozen are from other controllers and we can't safely combine them across controllers.